errors/middleware-upgrade-guide.md

```diff
@@ -0,0 +1,292 @@
+
+Previously, you could create a `_middleware.js` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:
+
+- Middleware in `pages/dashboard/_middleware.ts`
+- Middleware in `pages/dashboard/users/_middleware.ts`
+- A request to `/dashboard/users/*` **would match both.**
+
+Based on customer feedback, we have replaced this API with a single root Middleware.
+
+### How to upgrade
+
+You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory), and named without an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.
+
+Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:
+
+```typescript
+// middleware.ts
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  return NextResponse.rewrite(new URL('/about-2', request.url))
+}
+// config with custom matcher
+export const config = {
+  matcher: '/about/:path*',
+}
+```
+
+You can also use a conditional statement to only run the Middleware when it matches the `about/*` path:
+
+```typescript
+// <root>/middleware.js
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  if (request.nextUrl.pathname.startsWith('/about')) {
+    // This logic is only applied to /about
+  }
+
+  if (request.nextUrl.pathname.startsWith('/dashboard')) {
+    // This logic is only applied to /dashboard
+  }
+}
+```
+
+## No Response Body
+
+### Summary of changes
+
+- Middleware can no longer respond with a body
+- If your Middleware _does_ respond with a body, a runtime error will be thrown
+- Migrate to using `rewrites`/`redirects` to pages that handle authorization
+
+### Explanation
+
+Beta customers had explored using Middleware to handle authorization for their application. However, to ensure both the HTML and data payload (JSON file) are protected, we recommend checking authorization at the Page level.
+
+To help ensure security, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).
+
+The following patterns will no longer work:
+
+```js
+new Response('a text value')
+new Response(streamOrBuffer)
+new Response(JSON.stringify(obj), { headers: 'application/json' })
+```
+
+### How to upgrade
+
+For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrites`/`redirects` to Pages that show an authorization error, login forms, or to an API Route.
```

Comment:
The user action here is similar to https://github.com/vercel/next.js/pull/37382/files#diff-df4a93b5f2399985c81d6fb43adc2815bd39f190a0f770724bd7caf6a3628a73R292 for authoriztion. Is there some way to link them both together a bit more clearly?

Reply:
Done - added a link back at the bottom to this section